src/daemons/linuxd/src/lib.rs

```diff
@@ -536,50 +579,81 @@ impl<T: Sync + Send + 'static> LinuxDaemon<T> {
+    ) {
+        trace!("user_vm_reader_loop(): starting (uvm_id={uvm_id})");
+
+        loop {
+            match Self::recv(uvm_handle.get_user_vm_reader()).await {
+                Ok(message) => {
+                    trace!(
+                        "uservm.id={uvm_id}, message.source={:?}, message.destination={:?}, \
+                         message.type={:?}",
+                        { message.source },
+                        { message.destination },
+                        message.message_type,
+                    );
 
-        for uvm_id in ids {
-            let Some(uvm_handle) = user_vm_connections.get(&uvm_id).cloned() else {
-                continue;
-            };
-            let message: Message = match Self::recv(uvm_handle.get_user_vm_reader()).await {
-                Ok(m) => m,
-                Err(error_kind) => {
-                    // Log error message only if error was not caused by a disconnection.
-                    if error_kind != ErrorKind::ConnectionAborted
-                        && error_kind != ErrorKind::WouldBlock
+                    if uvm_events_tx
+                        .send(UserVmEvent::Message { uvm_id, message })
+                        .await
+                        .is_err()
                     {
-                        let reason: String =
-                            format!("failed to read message (error={error_kind:?})");
-                        error!("{reason}");
+                        // If the receiver side is gone, just exit.
+                        debug!(
+                            "user_vm_reader_loop(): dispatcher dropped receiver (uvm_id={uvm_id})"
+                        );
+                        break;
                     }
-                    return Err(uvm_id);
                 },
-            };
-
-            trace!(
-                "uservm.id={uvm_id}, message.source={:?}, message.destination={:?}, \
-                 message.type={:?}",
-                { message.source },
-                { message.destination },
-                message.message_type,
-            );
-
-            return Ok(Some((uvm_id, uvm_handle, message)));
+                Err(ErrorKind::ConnectionAborted) => {
+                    // User VM closed connection.
+                    trace!("user_vm_reader_loop(): connection aborted by peer (uvm_id={uvm_id})");
+
+                    // Ignore return value, as we are breaking anyway.
+                    let _ = uvm_events_tx
+                        .send(UserVmEvent::ConnectionClosed { uvm_id })
+                        .await;
+                    break;
+                },
+                Err(ErrorKind::WouldBlock) => {
+                    // Self::recv() properly handles partial reads by persisting them in a buffer,
+                    // so if it surfaces a WouldBlock we can safely just continue.
+                    // Optionally: tokio::task::yield_now().await;
+                    continue;
+                },
```

Comment:
I would not catch this and instead, let it trigger an error, as this is truly not expected and if it happens, it is an error.